tests/network/remote/remotes.c

```diff
@@ -28,30 +28,110 @@ void test_network_remote_remotes__cleanup(void)
 
-	cl_assert_equal_s(git_remote__urlfordirection(_remote, GIT_DIRECTION_FETCH),
-					  "git://github.com/libgit2/libgit2");
-	cl_assert_equal_s(git_remote__urlfordirection(_remote, GIT_DIRECTION_PUSH),
-					  "git://github.com/libgit2/libgit2");
+	cl_git_pass(git_remote__urlfordirection(&url, _remote, GIT_DIRECTION_FETCH, NULL));
+	cl_assert_equal_s(url, "git://github.com/libgit2/libgit2");
+
+	free(url);
+
+	cl_git_pass(git_remote__urlfordirection(&url, _remote, GIT_DIRECTION_PUSH, NULL));
+	cl_assert_equal_s(url, "git://github.com/libgit2/libgit2");
+
+	free(url);
 
 	cl_git_pass(git_remote_lookup(&_remote2, _repo, "test_with_pushurl"));
 	cl_assert_equal_s(git_remote_name(_remote2), "test_with_pushurl");
 	cl_assert_equal_s(git_remote_url(_remote2), "git://github.com/libgit2/fetchlibgit2");
 	cl_assert_equal_s(git_remote_pushurl(_remote2), "git://github.com/libgit2/pushlibgit2");
 
-	cl_assert_equal_s(git_remote__urlfordirection(_remote2, GIT_DIRECTION_FETCH),
-					  "git://github.com/libgit2/fetchlibgit2");
-	cl_assert_equal_s(git_remote__urlfordirection(_remote2, GIT_DIRECTION_PUSH),
-					  "git://github.com/libgit2/pushlibgit2");
+	cl_git_pass(git_remote__urlfordirection(&url, _remote2, GIT_DIRECTION_FETCH, NULL));
+	cl_assert_equal_s(url, "git://github.com/libgit2/fetchlibgit2");
+
+	free(url);
+
+	cl_git_pass(git_remote__urlfordirection(&url, _remote2, GIT_DIRECTION_PUSH, NULL));
+	cl_assert_equal_s(url, "git://github.com/libgit2/pushlibgit2");
+
+	free(url);
 
 	git_remote_free(_remote2);
 }
 
+static int test_network_remote_remotes__urlresolve_callback(char **url_resolved, const char *url, int direction, void *payload)
+{
+	const char *url_out;
+
+	cl_assert(git__strcmp(url, "git://github.com/libgit2/libgit2") == 0);
+	cl_assert(git__strcmp(payload, "payload") == 0);
+
+	url_out = (direction == GIT_DIRECTION_PUSH) ? "pushresolve" : "fetchresolve";
+
+	*url_resolved = git__strdup(url_out);
+
+	return GIT_OK;
+}
+
+void test_network_remote_remotes__urlresolve(void)
+{
+	char *url = NULL;
+
+	git_remote_callbacks callbacks = GIT_REMOTE_CALLBACKS_INIT;
+	callbacks.resolve_url = test_network_remote_remotes__urlresolve_callback;
+	callbacks.payload = "payload";
+
+	cl_assert_equal_s(git_remote_name(_remote), "test");
+	cl_assert_equal_s(git_remote_url(_remote), "git://github.com/libgit2/libgit2");
+	cl_assert(git_remote_pushurl(_remote) == NULL);
+
+	cl_git_pass(git_remote__urlfordirection(&url, _remote, GIT_DIRECTION_FETCH, &callbacks));
+	cl_assert_equal_s(url, "fetchresolve");
+
+	free(url);
+
+	cl_git_pass(git_remote__urlfordirection(&url, _remote, GIT_DIRECTION_PUSH, &callbacks));
+	cl_assert_equal_s(url, "pushresolve");
+
+	free(url);
+}
+
+static int test_network_remote_remotes__urlresolve_passthrough_callback(char **url_resolved, const char *url, int direction, void *payload)
+{
+	#pragma unused(url_resolved)
+	#pragma unused(url)
+	#pragma unused(direction)
+	#pragma unused(payload)
```

Comment:
Please use `GIT_UNUSED` instead of `#pragma`, which is not part of the C standard